4단계 - 구간 제거 기능 리뷰 요청

yomni opened the pull request:
안녕하세요! 리뷰어님!
이전에도 꼼꼼한 피드백으로 생각해볼 점을 많이 만들어 주셔서 감사했습니다. 😄
그 덕분에 이번 단계에서는 비교적 쉽게 구현이 가능했습니다!

이번 단계를 구현하다보니 몇 가지 궁금한점이 생겼는데요🧐

1. 저의 코드중에서 Section의 일급컬렉션인 Sections에 과도하게 많은 메서드가 집중되어 있다보니, 아무래도 뭔가 께름칙(?)한 느낌이 들어서 여쭤봅니다.   
  실무 차원에서 많은 역할을 가진 객체가 있는 것이 자연스러운 것인지, 만약 잘못된 것이라면 어떤식으로 회피(?)해 볼 수 있는 지 노하우가 궁금합니다!

2. 요구사항엔 없었지만, 동시성에 대한 이슈도 분명히 있을 것 같습니다. 해서, `List<Section>`의 인스턴스를 `java.util.concurrent.CopyOnWriteArrayList` 로 생성 했는데요, 동시성을 제어해본 경험이 없어서,  
- 동시성이 발생할 수도 있을 법한 케이스 찾기(동시에 같은 구간을 추가한다던지, 같은 노선에서 각각 다른 중간 구간을 삭제한다던지...)
- 동시성 문제에 대한 방어로 단순 ArrayList에서 CopyOnWriteArrayList 로 구현
이런 문제 인식 과정과 해결방법이 올바르게 되었는 지 궁금하네요!!

4단계 리뷰 미리 감사드립니다.
여전히 배울게 많다는 점에 있어서 설레네요!!

감사합니다!😁

yomni committed:
docs: 4단계 기능목록 정의

yomni committed:
feat: 종점을 제거하는 경우에 대한 구현 부분 추가

yomni committed:
feat: 상행 종점 노선, 중간 노선을 제거하는 경우에 대한 기능 구현

yomni committed:
feat: 예외케이스에 대한 구현 완료

- 구간이 하나인 노선에서 역을 제거하는 경우
- 노선에 없는 역을 제거하는 경우
- 동시성 대응을 위해 List<Section>을 CopyOnWriteArrayList로 정의

joojimin commented on src/test/java/nextstep/subway/utils/RestAssuredRequestBuilder.java:
```diff
@@ -51,6 +51,14 @@ public static ExtractableResponse<Response> delete(String path, Long id, HttpSta
                 .extract();
     }
 
+    public static ExtractableResponse<Response> delete(String path, String variable1, String variable2, HttpStatus status) {
```
공통코드 추출 좋네요!!

joojimin commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -165,4 +166,101 @@ void setUp() {
+     * When : 노선에 없는 역을 제거하려고 할 때,
+     * Then : 예외가 발생한다
+     */
+    @DisplayName("노선에 존재하지 않는 역을 제거하려고 했을 경우 에러")
```
에러가 발생하는 케이스도 잘 정리해주셧네요!

joojimin commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -2,23 +2,29 @@
     @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
     @JoinColumn(name = "line_id", foreignKey = @ForeignKey(name = "fk_section_of_line"))
-    private List<Section> sections = new ArrayList<>();
+    private List<Section> sections = new CopyOnWriteArrayList<>();
```
```
2. 요구사항엔 없었지만, 동시성에 대한 이슈도 분명히 있을 것 같습니다. 해서, List<Section>의 인스턴스를 java.util.concurrent.CopyOnWriteArrayList 로 생성 했는데요, 동시성을 제어해본 경험이 없어서,
* 동시성이 발생할 수도 있을 법한 케이스 찾기(동시에 같은 구간을 추가한다던지, 같은 노선에서 각각 다른 중간 구간을 삭제한다던지...)
* 동시성 문제에 대한 방어로 단순 ArrayList에서 CopyOnWriteArrayList 로 구현 이런 문제 인식 과정과 해결방법이 올바르게 되었는 지 궁금하네요!!
```
해당 질문의 답변은 여기서 드려야겠네요!!
음.. 먼저 왜 동시성의 문제가 생긴다고 생각하셧는지 궁금합니다!
해당 클래스는 Entity 내부에 들어가는 Embeddable로 설계된걸로 보이는데요!
Java의 쓰레드가 공유하는 클래스로 보이지는 않습니다! 그렇게 생각하신 이유를 좀 더 들어볼 수 있을까요~?

yomni replied:
안녕하세요!
우선 제가 처음 생각해본 동시성 문제에 대한 케이스는 다음과 같습니다.

![image](https://user-images.githubusercontent.com/45425050/204020747-a47108ea-9995-48d5-841d-db9fd1d06a9b.png)

만약 이런 경우 어떻게 될까? 라는 고민에서부터 시작된 문제였는데,
피드백 주신것을 토대로 검색하여 학습해본 결과,
1. `하나의 Request 당 thread pool 에서 한 개의 thread를 할당 받는다` 라는 점을 알 수 있었고,
2. 그렇다면, `Thread를 공유하는 클래스에서 동시성 문제가 발생할 수 있다`라고 생각해볼 수 있겠네요..

말씀하신 것처럼 제가 문제로 생각했던 클래스는
각 요청 thread 마다 별도로 인스턴스가 생성되는 클래스 이다보니 (**공유 멤버나 메서드가 존재하지 않음**)
동시성 문제를 고려하지 않아도 될 것 같습니다 😭

그렇다면, 추가로 궁금한 점이 있는데요 대규모의 대용량 서비스를 구축할 때
`동시성 문제에 대한 인지 과정` 이나 `동시성 테스트`는 실무 차원에서 어떤 방법들이 있는 지 궁금합니다!

joojimin replied:
`동시성 문제에 대한 인지 과정` 위에 말씀드린대로! 어플리케이션 단에서는 java의 쓰레드를 공유해서 사용하는 부분, 케이스가 있나를 찾아보고 그부분이 로직(코드)인지 데이터(DB)인지를 파악한 후 해결해나가고있습니다!

동시성 테스트는 실제 테스트코드로 java multi thread를 이용해 해당 상황을 만들거나 수많은 Request를 가정하여 stress test때 테스트해보는 걸로 진행해보고있어요!!

joojimin commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -2,23 +2,29 @@
+import java.util.concurrent.CopyOnWriteArrayList;
 
 @Embeddable
 public class Sections {
```
`1. 저의 코드중에서 Section의 일급컬렉션인 Sections에 과도하게 많은 메서드가 집중되어 있다보니, 아무래도 뭔가 께름칙(?)한 느낌이 들어서 여쭤봅니다. 실무 차원에서 많은 역할을 가진 객체가 있는 것이 자연스러운 것인지, 만약 잘못된 것이라면 어떤식으로 회피(?)해 볼 수 있는 지 노하우가 궁금합니다!`
=> 말씀하신 내용은 클래스 설계할때마다 항상 고민하는 부분인 것 같아요!
작업하다보면 한 클래스에 과도하게 많은 로직들이 들어가곤 하는데요! 그럴때마다 메서드들의 공통의 관심사는 어떤게 있는지 생각해보는 것 같습니다!
예를들어 유효성(validate) 체크, 객체 생성등 관심사가 같은 메서드들이 여러개있다면 다른 클래스를 생성하여 분리하고 있습니다!

joojimin reviewed (changes requested):
안녕하세요~ 용민님!
리뷰가 늦어서 죄송합니다!
질문해주신 부분은 같이 짚고 넘어가면 좋을거 같네요!
코멘트 한번 확인부탁드립니다!

yomni committed:
refactor: List<Section>의 구현체를 CopyOnWriteArrayList --> ArrayList로 복원

동시성 문제의 발현 조건에 대한 인지 부족으로 발생하지도 않을 문제에 대해 생각해본 듯.. Spring Web에서의 request 처리는 thread pool 에서 request 당 1개의 thread를 할당하게 되고, 이는 thread 차원에서 공유되는 자원이 아니라면 동시성 문제에 대해 고려할 사항이 아님

joojimin reviewed (approved):
안녕하세요~ 용민님!
피드백반영 고생하셧습니다! 
미션진행하는동안 많은 고민을 하신것같아서 감명깊었습니다!
ㅎㅎ 남은미션도 화이팅입니다!
